### DISCUSSION TITLE
How should we deal with abandonware

### DISCUSSION THREAD

**Discussion by muhlemmer** [zitadel/zitadel#5258] [Ideas]
Like many other projects Zitadel and some of its libraries, like [oidc](https://github.com/zitadel/oidc) currently make use of multiple open source packages. This helps keeping development cost low by reusing widely needed functionality. The downside is that we depend on the maintainers of those packages to keep them up to date and more importantly deal with security updates. It does happen that development stalls on certain project. Usually  because there is no more incentive my its maintainer:

- The company behind a project ceases to exist;
- The tech implemented by a package is no longer relevant for the developer;
- A package was written as free-time project or potential side-hustle and never set off;

Some examples of "abandonware" that we are aware off and need to move away from in zitadel are the [gorilla toolkit](https://github.com/gorilla#gorilla-toolkit) (`mux`, `schema`) and [square/go-jose](https://github.com/square/go-jose).

Because we take management of dependencies (and the resulting quality of our product) serious at Zitadel  we want to get a discussion going on a few items:

1. How can we monitor if a library became abandonware? For example, [dependabot](https://github.com/dependabot)  helps us when our dependencies are outdated, but doesn't warn us if a dependency became inactive. Does anyone now about tooling that monitor inactivity of a repository?
2.  What are decent alternatives for currently abandoned dependencies? (Lower refactor cost vs. stability of replacement);
3. Perhaps it's better to adopt critical dependencies, like [square/go-jose](https://github.com/square/go-jose). We would have to establish if it's cheaper to maintain than refactor existing code.
4. Consider reaching out to other companies with the same problems with the same libraries and somehow cooperate in bringing new life into them. (Donations, perhaps?)

We mean to have this discussion with our own team as well with the community. Any ideas or experiences are more than welcome!

**Comment by muhlemmer** [comment]
For `gorilla/mux` and `gorilla/schema` there seems to be valid alternatives: `go-chi/chi` and `go-chi/render`. https://github.com/zitadel/zitadel/issues/4285#issuecomment-1449890898

However, and alternative for `gorilla/securecookie` must still be found. I've come across https://github.com/chmike/securecookie just now, which seems to be a reasonable alternative. Code looks clean and maintained. Any other opinions?

**Comment by fforootd** [comment]
> 1. How can we monitor if a library became abandonware? For example, [dependabot](https://github.com/dependabot) helps us when our dependencies are outdated, but doesn't warn us if a dependency became inactive. Does anyone now about tooling that monitor inactivity of a repository?

Well the only thing i found that worked halfway is this list from snyk https://snyk.io/advisor/packages/golang/maintenance/inactive

![image](https://user-images.githubusercontent.com/9879976/222276601-e48b31e1-0818-4bdf-b1bd-238bfbf23df3.png)

> 2. What are decent alternatives for currently abandoned dependencies? (Lower refactor cost vs. stability of replacement);

If we have drop-in replacements I would favor that. But for example in the case of go-jose (square and go-jose) I would opt for a "copy and own" strategy since it is a crucial piece of code. We can approach the community to participate in their effort but it might be hard to gain a foothold with the maintainers.
Btw. both a flagged as inactive with snyk 😁 

> 3. Perhaps it's better to adopt critical dependencies, like [square/go-jose](https://github.com/square/go-jose). We would have to establish if it's cheaper to maintain than refactor existing code.

Agreed as above

> 4. Consider reaching out to other companies with the same problems with the same libraries and somehow cooperate in bringing new life into them. (Donations, perhaps?)

Well, we can try and approach other companies, what comes to mind is authentik, ory, casdoor, ..., who rely on jose for example.
But I am unsure about this, companies can be to self centered sometime 😁 but it might be worth a shot.
